[Federation] Setting up CoreDNS as DNS provider for Cluster Federation

shashidharatd opened the pull request:
Addresses and Fixes: #2197

cc @madhusudancs @kubernetes/sig-federation-misc

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2810)
<!-- Reviewable:end -->

madhusudancs commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 28 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GiWmBNTL8uz9QQqQ:-Kg1GiWmBNTL8uz9QQqR:b-ho6r9i) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L28)):*
> ```Markdown
> 
> To deploy CoreDNS we shall make use of helm charts. CoreDNS will be
> deployed with Etcd as the backend and should be pre-installed. Etcd
> ```

`etcd` is usually spelled with a lower case `e`. Also, please link to their website - https://coreos.com/etcd

Same comments for all other occurrences. 

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GxNTPs7Scjm4hUvd:-Kg1GxNTPs7Scjm4hUve:b-6o232u) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L33)):*
> ```Markdown
> 
> ```shell
> $ helm install --name etcd-operator --namespace ns stable/etcd-operator
> ```

Can we use a better, more descriptive namespace?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 41 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GlW-UkzzGrroldFH:-Kg1GlW-UkzzGrroldFI:b-3t6k18) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L41)):*
> ```Markdown
> 
> After deployment succeeds, Etcd can be accessed with
> "http://etcd-cluster:2379" endpoint within the host cluster.
> ```

Remove the quotes and just linkify the URL.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 44 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1H94Is0aatsxs7J5i:-Kg1H94Is0aatsxs7J5j:bfxldb2) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L44)):*
> ```Markdown
> 
> CoreDNS default configuration should be customized to suit federation
> and shown below is the Values.yaml, which overrides the default
> ```

Link to the default `Values.yaml` file?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HImQy93kla48Wvx5:-Kg1HImQy93kla48Wvx6:b-p8ck88) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
> ```Markdown
>  - `isClusterService` specifies whether CoreDNS should be deployed as
> cluster-service, which is default. You need to set it to false, so that
> CoreDNS is deployed as kubernetes application service.
> ```

What's the difference between the two types? I am not sure I know the distinction.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 70 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HdZKKLisPOwu3E_n:-Kg1HdZKKLisPOwu3E_o:b-y3kv47) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L70)):*
> ```Markdown
> cluster.
>  - Disable `middleware.kubernetes`, which is enabled by default by
> setting `middleware.kubernetes.enabled` to false.
> ```

Do `CoreDNS` users understand this configuration? I don't know much about CoreDNS and I can't say whether I want this or not.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 72 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HoREqF3NDKtzgRpv:-Kg1HoREqF3NDKtzgRpw:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L72)):*
> ```Markdown
> setting `middleware.kubernetes.enabled` to false.
>  - Enable `middleware.etcd` by setting `middleware.etcd.enabled` to
> true.
> ```

Same as above.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 89 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1ILzEmG9jvmDGdbOV:-Kg1ILzEmG9jvmDGdbOW:b-cpqdun) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L89)):*
> ```Markdown
> 
> Federation control plane can be deployed using `kubefed init`. CoreDNS
> can be chosen as DNS provider by specifying two additional parameters.
> ```

It is probably worth mentioning that `middleware.etcd.zones` and `--dns-zone-name` flag to `kubefed init` should match. 

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 98 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1JZHS8mBRr8VvMDjX:-Kg1JZHT90ircXSfu8Oo:b-j4nyjf) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L98)):*
> ```Markdown
> ```shell
> [Global]
> etcd-endpoints = http://etcd-cluster.ns:2379
> ```

I would make the host name part consistent. I understand why you are doing this. But I think it is worth making it consistent everywhere. Using `etcd-cluster.ns` everywhere isn't harmful right?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 99 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1JGg-45osFAkGFblD:-Kg1JGg-45osFAkGFblE:b-l3frtq) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L99)):*
> ```Markdown
> [Global]
> etcd-endpoints = http://etcd-cluster.ns:2379
> zones = example.com
> ```

Trailing dot for consistency as we did here - https://github.com/kubernetes/kubernetes.github.io/pull/2899#pullrequestreview-27716728

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1Jx-Lwtjtcv60V1qT:-Kg1Jx-M6pJc-BybXEEh:b-myfyfi) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L103)):*
> ```Markdown
> 
>  - `etcd-endpoints` is the endpoint to access Etcd.
>  - `zones` is the federation domain for which CoreDNS is authoritative.
> ```

and say "which is same as `--dns-zone-name`"

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 105 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1KCawNvqTxoPIrHu2:-Kg1KCawNvqTxoPIrHu3:bail156) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L105)):*
> ```Markdown
>  - `zones` is the federation domain for which CoreDNS is authoritative.
> 
> ## Setup CoreDNS server in nameserver resolv chain
> ```

resolv.conf chain?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 109 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1KIgSXTkSk90dHrsg:-Kg1KIgSXTkSk90dHrsh:bail156) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L109)):*
> ```Markdown
> Once the federation control plane is deployed and federated clusters
> are joined to the federation, you need to add the CoreDNS server to
> pod's nameserver resolv chain in all the federated clusters as this
> ```

resolv.conf chain?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 115 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1O3DrIT6Bgl-AKswi:-Kg1O3DrIT6Bgl-AKswj:b-j5n266) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L115)):*
> ```Markdown
> 
> ```shell
>         - --server=/example.com/<CoreDNS endpoint>
> ```

Is it not possible to use the `kube-dns` ConfigMap for this yet?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, Handled review comments. PTAL

---

Review status: 1 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 28 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GiWmBNTL8uz9QQqQ:-Kg1rBZosixtxVgtfNvF:b-1r0mvf) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L28)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

`etcd` is usually spelled with a lower case `e`. Also, please link to their website - https://coreos.com/etcd

Same comments for all other occurrences. 
</blockquote></details>

ok. Done

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GxNTPs7Scjm4hUvd:-Kg1rGMaqlQ2PMj4gZe9:b-yhk8c8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L33)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can we use a better, more descriptive namespace?
</blockquote></details>

used my-namespace instead of ns

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 41 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GlW-UkzzGrroldFH:-Kg1rNBwAu8146woYt2-:byl1h1n) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L41)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Remove the quotes and just linkify the URL.
</blockquote></details>

removed the quotes

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 44 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1H94Is0aatsxs7J5i:-Kg206wIg9jxmzmKUvE0:bij6fgk) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L44)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Link to the default `Values.yaml` file?
</blockquote></details>

CoreDNS chart is not merged yet, so the link would not be stable if i link it to PR. On the other hand Values.yaml in helm charts is well known, so i think user should not have any difficulty finding out the default configuration given the chart.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HImQy93kla48Wvx5:-Kg1rvMKjMGdzBiCOdla:b1atj5m) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

What's the difference between the two types? I am not sure I know the distinction.
</blockquote></details>

CoreDNS chart can be deployed basically as 2 variants.
 - if isClusterService is true, then CoreDNS is deployed similar to kube-dns addon. 
 - if isClusterService is false, then CoreDNS is deployed as any other k8s app. 

they basically differ in labels and annotations.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 70 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HdZKKLisPOwu3E_n:-Kg1thbLqHPuK5oG1kSW:b-yctpgg) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L70)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Do `CoreDNS` users understand this configuration? I don't know much about CoreDNS and I can't say whether I want this or not.
</blockquote></details>

CoreDNS users do understand about the middleware. I had given the reference Values.yaml above which can be used to deploy CoreDNS for federation dns server. This explanation is just to let know the user of the configuration details and i believe whoever wants to tweak the configuration would know how to fit it to their needs.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 72 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HoREqF3NDKtzgRpv:-Kg20hh658wxqHWAfDNB:b1u2wre) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L72)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Same as previous reply

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 89 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1ILzEmG9jvmDGdbOV:-Kg20pI6qNs1iBUcd3qz:bkd29fa) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L89)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

It is probably worth mentioning that `middleware.etcd.zones` and `--dns-zone-name` flag to `kubefed init` should match. 
</blockquote></details>

Added a Note

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 98 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1JZHS8mBRr8VvMDjX:-Kg1yMUfCRsU7E_CsutI:bjhc6cp) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L98)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would make the host name part consistent. I understand why you are doing this. But I think it is worth making it consistent everywhere. Using `etcd-cluster.ns` everywhere isn't harmful right?
</blockquote></details>

Agree, used etcd-cluster.my-namespace everywhere

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 99 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1JGg-45osFAkGFblD:-Kg1ykLfDsVvWgACy3f8:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L99)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Trailing dot for consistency as we did here - https://github.com/kubernetes/kubernetes.github.io/pull/2899#pullrequestreview-27716728
</blockquote></details>

Done

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1Jx-Lwtjtcv60V1qT:-Kg1yn6i0a0IiRDWUwLT:bqmjwcz) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L103)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

and say "which is same as `--dns-zone-name`"
</blockquote></details>

Added.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 105 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1KCawNvqTxoPIrHu2:-Kg1zi0-9c--lha38wUU:bavwohv) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L105)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

resolv.conf chain?
</blockquote></details>

Changed. Done

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 109 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1KIgSXTkSk90dHrsg:-Kg1zkMvlAzdG-4vdZG0:bavwohv) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L109)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

resolv.conf chain?
</blockquote></details>

Changed. Done

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 115 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1O3DrIT6Bgl-AKswi:-Kg1zn8CbR8h1opXPZk4:b3l5tzq) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L115)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Is it not possible to use the `kube-dns` ConfigMap for this yet?
</blockquote></details>

kube-dns ConfigMap in the current release cannot handle nameserver listening on any port other than 53. I did a PR to specify port for nameserver, but that would be included in subsequent releases. 
If CoreDNS is deployed as NodePort service. than it will be accessible on some random port which cannot be configured using kube-dns ConfigMap method.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-:-Kg277Be7P_kPziB9RHD:b-rcasxk)*
<!-- Sent from Reviewable.io -->

csbell commented:
Madhu -- can you lgtm the changes if they look good to you?

shashidharatd commented:
@madhusudancs, the CoreDNS helm chart PR is merged now. request a LGTM from you. is the fixes ok?

clenimar commented:
*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 99 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KgyXEhY_P3YietczNXL:-KgyXEhY_P3YietczNXM:bvtwroc) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/4f8d1f1c4ccb0271c15855e4040f9de8d5cd7d69/docs/tutorials/federation/set-up-coredns-provider-federation.md#L99)):*
> ```Markdown
> [Global]
> etcd-endpoints = http://etcd-cluster.my-namespace:2379
> zones = example.com.
> ```

Does the user need to own this domain name? (In the light of https://github.com/kubernetes/kubernetes/issues/43599)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@shashidharatd a few minor comments.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GxNTPs7Scjm4hUvd:-Kh-87ALE8Lj8iHbiQd3:brzoavm) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L33)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

used my-namespace instead of ns
</blockquote></details>

I would have preferred a descriptive one. Something like `dns` or `coredns`, but this is good enough.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 41 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GlW-UkzzGrroldFH:-Kh-8QjOD7y7_eZl4BMA:bi99q5r) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L41)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

removed the quotes
</blockquote></details>

Can you please linkify this so that people following the doc can just click the link? `[http://etcd-cluster.my-namespace:2379](http://etcd-cluster.my-namespace:2379)`

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 44 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1H94Is0aatsxs7J5i:-Kh-8eoEi8SKtPuupiKG:b-qezkjt) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L44)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

CoreDNS chart is not merged yet, so the link would not be stable if i link it to PR. On the other hand Values.yaml in helm charts is well known, so i think user should not have any difficulty finding out the default configuration given the chart.
</blockquote></details>

Ok, fair enough.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HImQy93kla48Wvx5:-Kh-8s-UYEKiaGH_Nz9Q:b-9gzgvk) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

CoreDNS chart can be deployed basically as 2 variants.
 - if isClusterService is true, then CoreDNS is deployed similar to kube-dns addon. 
 - if isClusterService is false, then CoreDNS is deployed as any other k8s app. 

they basically differ in labels and annotations.
</blockquote></details>

Oh I see what you mean. This defines whether the deployment is a cluster component. Is that right? Could you please clarify this in the doc. It is not clear right now.

If that's the case, should users care about it? What are the pros and cons?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 70 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HdZKKLisPOwu3E_n:-Kh-9MtrZfoLR51Q26_J:bl0xfe8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L70)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

CoreDNS users do understand about the middleware. I had given the reference Values.yaml above which can be used to deploy CoreDNS for federation dns server. This explanation is just to let know the user of the configuration details and i believe whoever wants to tweak the configuration would know how to fit it to their needs.
</blockquote></details>

Ok, cool.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 72 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HoREqF3NDKtzgRpv:-Kh-9PgKsFDMqs4WKEzI:bb6sh2i) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L72)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Same as previous reply
</blockquote></details>

Ack.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 115 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1O3DrIT6Bgl-AKswi:-Kh-B4UMNvi3CJIndUTZ:bm2tnrb) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L115)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

kube-dns ConfigMap in the current release cannot handle nameserver listening on any port other than 53. I did a PR to specify port for nameserver, but that would be included in subsequent releases. 
If CoreDNS is deployed as NodePort service. than it will be accessible on some random port which cannot be configured using kube-dns ConfigMap method.

</blockquote></details>

I see. Do you think it is worth documenting both these techniques along with your explanation above?

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 99 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KgyXEhY_P3YietczNXL:-Kh-9ojqgWhUUDn5rWnn:b-z20akm) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/4f8d1f1c4ccb0271c15855e4040f9de8d5cd7d69/docs/tutorials/federation/set-up-coredns-provider-federation.md#L99)):*
<details><summary><i>Previously, clenimar (Clenimar Filemon) wrote…</i></summary><blockquote>

Does the user need to own this domain name? (In the light of https://github.com/kubernetes/kubernetes/issues/43599)
</blockquote></details>

Not necessary if you are using CoreDNS as the DNS provider. However, it is the user's (read: cluster operator's) responsibility to
1. Appropriately configure cluster's `kube-dns` to forward all requests to `example.com.` to CoreDNS.
2. Make CoreDNS the authoritative server for `example.com.`
3. Ensure external clients reach CoreDNS for the DNS queries for `example.com.`.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-:-Kh-BVqMA_ccepG0dPaK:b-sxzknz)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, have tried adressing your comments. PTAL

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GxNTPs7Scjm4hUvd:-Kh0J14R4xNzHxBUVI7X:bxhzo50) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L33)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would have preferred a descriptive one. Something like `dns` or `coredns`, but this is good enough.
</blockquote></details>

Using something like `coredns` might have confused, since the keyword is used for other purposes also.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 41 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GlW-UkzzGrroldFH:-Kh0JYqbZOsHKgRefs0p:b-y3cv5t) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L41)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can you please linkify this so that people following the doc can just click the link? `[http://etcd-cluster.my-namespace:2379](http://etcd-cluster.my-namespace:2379)`
</blockquote></details>

oh, i misunderstood the comment. Fixed Now.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HImQy93kla48Wvx5:-Kh0i6UkwSvzbK6tqxS8:b6ryuzo) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Oh I see what you mean. This defines whether the deployment is a cluster component. Is that right? Could you please clarify this in the doc. It is not clear right now.

If that's the case, should users care about it? What are the pros and cons?
</blockquote></details>

For federation use case. CoreDNS has to be deployed as normal k8s app and this is not default while installing helm chart. So i have given the overriding configuration to deploy CoreDNS as a normal k8s app with etcd as middleware.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 115 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1O3DrIT6Bgl-AKswi:-Kh0im_-N0JEaPY54EJ7:b12s7wj) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L115)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I see. Do you think it is worth documenting both these techniques along with your explanation above?
</blockquote></details>

Once the kube-dns configuration related PR is merged and available to users. users need not do this manual configuration at all. we should remove this section in future. So i feel we should leave out those details which we do take care of.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-:-Kh0jxrI_UkbAUM91tae:b-a1n5wv)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1GxNTPs7Scjm4hUvd:-KhDsvnrjyjQYAykI7v7:b-gxztsp) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L33)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Using something like `coredns` might have confused, since the keyword is used for other purposes also.
</blockquote></details>

Fair enough.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1HImQy93kla48Wvx5:-KhDtHBlHSdMszDVNxjw:b-mtgy2u) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

For federation use case. CoreDNS has to be deployed as normal k8s app and this is not default while installing helm chart. So i have given the overriding configuration to deploy CoreDNS as a normal k8s app with etcd as middleware.
</blockquote></details>

Ok makes sense.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 115 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-Kg1O3DrIT6Bgl-AKswi:-KhDt_v5ekRBX6MBQoPj:b2nhll9) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/d45335bcd71743f2094169b9664305703cc35819/docs/tutorials/federation/set-up-coredns-provider-federation.md#L115)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Once the kube-dns configuration related PR is merged and available to users. users need not do this manual configuration at all. we should remove this section in future. So i feel we should leave out those details which we do take care of.
</blockquote></details>

Ok.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@devin-donnelly @chenopis @jaredbhatti could you please take a look?

madhusudancs commented:
@shashidharatd this is also needs a rebase.

shashidharatd commented:
@madhusudancs, sorry for the delay. rebased now

arthur0 commented:
Hello @shashidharatd, I tried `helm install --name etcd-operator stable/etcd-operator` (with and without the `--namespace` flag) unsuccessfully. Do you know any reason for that?
```
Error: release etcd-operator failed: the server does not allow access to the requested resource (get namespaces default)
```

shashidharatd commented:
Hello @arthur0,
  I am able to install the chart with/without namespace successfully. i am suspecting you are having issue with RBAC settings in your k8s cluster.
Anyway it would be good to post this issue in https://github.com/kubernetes/charts instead of here. there would be more users/developers to help out.

csbell commented:
Another possibility: federation doesn't create a _default_ namespace. The failing "gets" suggest that might be the case.

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+{:toc}
+
+This guide explains how to configure and deploy CoreDNS to be used as
+DNS provider for Cluster Federation.
```
add "the" before "DNS provider..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+- shashidharatd
+title: Setting up CoreDNS as DNS provider for Cluster Federation
+---
+
```
If you could please use the [Tutorial Template](https://kubernetes.io/docs/home/contribute/page-templates/#tutorial_template), that would really help us out. Thanks!

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+This guide explains how to configure and deploy CoreDNS to be used as
+DNS provider for Cluster Federation.
+
+CoreDNS can be deployed in various configuration suiting the cluster
```
"configuration" -> "configurations"

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+DNS provider for Cluster Federation.
+
+CoreDNS can be deployed in various configuration suiting the cluster
+federation deployment. In this guide we are explaining CoreDNS
```
The sentence "In this guide..." is redundant and can be removed because the points it makes are already mentioned above.

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+## Deploying CoreDNS and etcd charts
+
+To deploy CoreDNS we shall make use of helm charts. CoreDNS will be
```
add comma: "To deploy CoreDNS`,` we shall..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+To deploy CoreDNS we shall make use of helm charts. CoreDNS will be
+deployed with [etcd](https://coreos.com/etcd) as the backend and should
+be pre-installed. etcd also can be deployed using helm charts. Shown
```
"etcd also can be deployed..." -> "etct can also be deployed..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+> Note: etcd default deployment configurations can be overridden, suiting the
+host cluster.
+
+After deployment succeeds, etcd can be accessed with
```
add "the": "...can be accessed with..." -> "...can be accessed with `the`..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+After deployment succeeds, etcd can be accessed with
+[http://etcd-cluster.my-namespace:2379](http://etcd-cluster.my-namespace:2379) endpoint within the host cluster.
+
+CoreDNS default configuration should be customized to suit federation
```
add "the": "`The` CoreDNS default..." and "...customized to suit `the` federation."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+[http://etcd-cluster.my-namespace:2379](http://etcd-cluster.my-namespace:2379) endpoint within the host cluster.
+
+CoreDNS default configuration should be customized to suit federation
+and shown below is the Values.yaml, which overrides the default
```
start a new sentence: "...suit the federation.` `Shown below is..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+CoreDNS default configuration should be customized to suit federation
+and shown below is the Values.yaml, which overrides the default
+configuration parameters on CoreDNS chart.
```
add "the": "...on `the` CoreDNS chart."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+The above configuration file needs some explanation:
+
+ - `isClusterService` specifies whether CoreDNS should be deployed as
+cluster-service, which is default. You need to set it to false, so that
```
add "a" and "the": "`a` cluster-service, which is `the` default."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+ - `isClusterService` specifies whether CoreDNS should be deployed as
+cluster-service, which is default. You need to set it to false, so that
+CoreDNS is deployed as kubernetes application service.
```
add "a," capitalize Kubernetes: "...deployed as `a Kubernetes` application service."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+ - `isClusterService` specifies whether CoreDNS should be deployed as
+cluster-service, which is default. You need to set it to false, so that
+CoreDNS is deployed as kubernetes application service.
+ - `serviceType` specifies the type of kubernetes service to be created
```
capitalize Kubernetes everywhere

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+cluster-service, which is default. You need to set it to false, so that
+CoreDNS is deployed as kubernetes application service.
+ - `serviceType` specifies the type of kubernetes service to be created
+for CoreDNS. You need to choose either "LoadBalancer" or "NodePort" in
```
Suggested wording: 

You need to choose either "LoadBalancer" or "NodePort" to make the CoreDNS service accessible outside the Kubernetes cluster.

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+## Deploying Federation with CoreDNS as DNS provider
+
+Federation control plane can be deployed using `kubefed init`. CoreDNS
```
add "the": "`The` Federation control plane..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+## Deploying Federation with CoreDNS as DNS provider
+
+Federation control plane can be deployed using `kubefed init`. CoreDNS
+can be chosen as DNS provider by specifying two additional parameters.
```
add "the": "...can be chosen as `the` DNS provider..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+Once the federation control plane is deployed and federated clusters
+are joined to the federation, you need to add the CoreDNS server to
+pod's nameserver resolv.conf chain in all the federated clusters as this
```
add "the": "...`the` pod's nameserver..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+
+Replace `example.com` above with federation domain.
+
+> Note: Adding CoreDNS server to pod's nameserver resolv.conf chain will be
```
add "the": "...server to `the` pod's nameserver..."

chenopis commented on docs/tutorials/federation/set-up-coredns-provider-federation.md:
```diff
@@ -0,0 +1,126 @@
+automated in subsequent releases.
+
+
+Now the federated cluster is ready for cross-cluster service dicovery !
```
typo and remove extra space: "...cross-cluster service `discovery!`"

shashidharatd commented:
@chenopis, i have addressed your comments. PTAL

---

Review status: 1 of 3 files reviewed at latest revision, 21 unresolved discussions.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 6 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-6:-Kha83_VuAmudwxz2AJx:blfqi93) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L6)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

If you could please use the [Tutorial Template](https://kubernetes.io/docs/home/contribute/page-templates/#tutorial_template), that would really help us out. Thanks!
</blockquote></details>

I have used the template now. PTAL

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 11 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-11:-Kha86lJ6yILRI6R6IPG:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L11)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the" before "DNS provider..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 13 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-13:-Kha89bxrpl6fXZ9LdaG:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L13)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

"configuration" -> "configurations"
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 14 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-14:-Kha8IH7VNmLjM7923Wa:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L14)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

The sentence "In this guide..." is redundant and can be removed because the points it makes are already mentioned above.
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 27 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-27:-Kha8LC36vZMW4PPnhVs:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L27)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add comma: "To deploy CoreDNS`,` we shall..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 29 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-29:-Kha8MhKH_wmnEPiTPvD:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L29)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

"etcd also can be deployed..." -> "etct can also be deployed..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 40 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-40:-Kha8RAi-qHDhuhTNP-y:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L40)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "...can be accessed with..." -> "...can be accessed with `the`..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 43 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-43:-Kha8TSIF_HYz8eKPaBf:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L43)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "`The` CoreDNS default..." and "...customized to suit `the` federation."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 44 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-44:-Kha8UI1mtF6DQllGfhF:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L44)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

start a new sentence: "...suit the federation.` `Shown below is..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 45 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-45:-Kha8VeBgzI7cE1JwhJc:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L45)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "...on `the` CoreDNS chart."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 63 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-63:-Kha8Wm_LBX-WtM9LnCj:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L63)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "a" and "the": "`a` cluster-service, which is `the` default."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 64 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-64:-Kha8XhIpByEwQIJw94G:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L64)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "a," capitalize Kubernetes: "...deployed as `a Kubernetes` application service."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 65 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-65:-Kha8YihG3-iwcEQkLZ9:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L65)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

capitalize Kubernetes everywhere
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 66 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-66:-Kha8_jIoeEnhSGVqTKw:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L66)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

Suggested wording: 

You need to choose either "LoadBalancer" or "NodePort" to make the CoreDNS service accessible outside the Kubernetes cluster.
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 88 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-88:-Kha8arsNsB5wzScIT5G:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L88)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "`The` Federation control plane..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 89 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-89:-Kha8bvnuggB-sHjcS4e:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L89)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "...can be chosen as `the` DNS provider..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 111 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-111:-Kha8dPaoc8E40vhvR7e:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L111)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "...`the` pod's nameserver..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 122 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-122:-Kha8edUUlgcLE305B92:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L122)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

add "the": "...server to `the` pod's nameserver..."
</blockquote></details>

Done.

---

*[docs/tutorials/federation/set-up-coredns-provider-federation.md, line 126 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-KfD9E4Xkyh3VSFXBw6J-r3-126:-Kha8flAWjfTa_IL3NuZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/c3c480fdf83ec973a5eee612f9d1c40f2e6aff8a/docs/tutorials/federation/set-up-coredns-provider-federation.md#L126)):*
<details><summary><i>Previously, chenopis (Andrew Chen) wrote…</i></summary><blockquote>

typo and remove extra space: "...cross-cluster service `discovery!`"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes.github.io/2810#-:-Kha8iS2DVam0T-gv3G6:bshjkik)*
<!-- Sent from Reviewable.io -->

chenopis commented:
/lgtm @shashidharatd @madhusudancs Anything else before I merge this?

shashidharatd commented:
loogs good to me. @madhusudancs PTAL one final time.
Thanks

madhusudancs commented:
@chenopis @shashidharatd looks good to me too. Go go go!

shashidharatd commented:
@chenopis, @madhusudancs, This document should be available for 1.6, should i raise to release-1.6 branch also?

chrisjones262 commented:
Note that using ServiceType = "LoadBalancer"  gives this result:
Error: release coredns failed: Service "coredns-coredns" is invalid: spec.ports: Invalid value: [{"name":"dns","protocol":"UDP","port":53,"targetPort":53,"nodePort":0},
{"name":"dns-tcp","protocol":"TCP","port":53,"targetPort":53,"nodePort":0}]: cannot create an external load balancer with mix protocols    

Using ServiceType = "NodePort" seems to be necessary if you want to support both TCP and UDP.

shashidharatd commented:
@chrisjones262, on what cloud provider are you seeing the above issue?
Request you to file an issue in https://github.com/kubernetes/charts
AFAIK, there seems to be a bug in GCE loadbalancer, which does not allow to create a loadbalancer with same port for UDP & TCP. Are you by chance using GCE?
I hope this problem does not exist on other cloud providers like AWS or openstack

chrisjones262 commented:
It is on Azure Container Service.   I would guess it’s cloud-provider independent.

From: Shashidhara T D [mailto:notifications@github.com]
Sent: Friday, April 14, 2017 10:34 AM
To: kubernetes/kubernetes.github.io <kubernetes.github.io@noreply.github.com>
Cc: Christopher Jones (MCS) <cjones@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes.github.io] [Federation] Setting up CoreDNS as DNS provider for Cluster Federation (#2810)


@chrisjones262<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fchrisjones262&data=02%7C01%7Ccjones%40microsoft.com%7Cc5b3ceb90d854dd39a8008d483434911%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636277772391055382&sdata=hmgRRv3WKeKFQtyo9nTGucRUGvCUFJRIAxpBg6nCNMY%3D&reserved=0>, on what cloud provider are you seeing the above issue?
Request you to file an issue in https://github.com/kubernetes/charts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fcharts&data=02%7C01%7Ccjones%40microsoft.com%7Cc5b3ceb90d854dd39a8008d483434911%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636277772391055382&sdata=wcWxMsRhQQ%2FsPVzf7iH1%2FFOfqriIgI0isex4s3zKDAc%3D&reserved=0>
AFAIK, there seems to be a bug in GCE loadbalancer, which does not allow to create a loadbalancer with same port for UDP & TCP. Are you by chance using GCE?
I hope this problem does not exist on other cloud providers like AWS or openstack

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fkubernetes.github.io%2Fpull%2F2810%23issuecomment-294164044&data=02%7C01%7Ccjones%40microsoft.com%7Cc5b3ceb90d854dd39a8008d483434911%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636277772391055382&sdata=dhpOmzV9W54mH%2BR4fn9lpdRJa%2F%2Fslet6UiOfUxzE%2FwU%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJt1ZZBA8H9d_2JC_1diaT_9VYID6YsFks5rv4PUgaJpZM4MdAO2&data=02%7C01%7Ccjones%40microsoft.com%7Cc5b3ceb90d854dd39a8008d483434911%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636277772391055382&sdata=eI22HjmNMoXcJVwP8qdHM4Ro%2BgiLnapiF90fNb2tth0%3D&reserved=0>.

chenopis commented:
@shashidharatd Since 1.6 was already released, I can merge this directly into `master`. Once you think the issue w/ @chrisjones262 is sufficiently resolved, I will merge this. Cheers

shashidharatd commented:
@chrisjones262, interesting to know that loadbalancers in 'Azure Container Service' also has similar characteristics as that GCE loadbalancers. We can just start the service to listen either on TCP or UDP in that case and think about implications for DNS server. I would request you to file this issue in https://github.com/kubernetes/charts as the fix need to go in that repo for CoreDNS helm chart.

shashidharatd commented:
@chenopis, just curious, can this documentation be part of 1.6.2 patch release? Does the hosted https://kubernetes.io reflect the changes done to patch release?
I wish that this documentation is merged and visible to greater audience ASAP, if it has to go in 1.7 release then we need to wait for another quarter.

Also the issue raised by @chrisjones262, affect subset of environments. IMHO, we should not hold back this documentation pr for that issue as there is no change required here. that issue is being addressed in https://github.com/kubernetes/charts/pull/928 and would take its own time to merge and is independent of kubernetes release cycles. @chrisjones262 do you agree?

chenopis commented:
@shashidharatd As long as that issue is being addressed, I'm fine w/ merging this now. Thx

chrisjones262 commented:
Sorry for the delay.  I agree.

From: Shashidhara T D [mailto:notifications@github.com]
Sent: Sunday, April 16, 2017 11:58 PM
To: kubernetes/kubernetes.github.io <kubernetes.github.io@noreply.github.com>
Cc: Christopher Jones (MCS) <cjones@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [kubernetes/kubernetes.github.io] [Federation] Setting up CoreDNS as DNS provider for Cluster Federation (#2810)


@chenopis<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fchenopis&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=gkiMGeqhYtlTm53yYH%2FCPAkTBp2MKHx2kPlSWT1YiHE%3D&reserved=0>, just curious, can this documentation be part of 1.6.2 patch release? Does the hosted https://kubernetes.io<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fkubernetes.io&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=4fTbdnMdmm%2BI0xZ2o%2BEZK4CJDR99%2F9SnBBGrT4Sqgpo%3D&reserved=0> reflect the changes done to patch release?
I wish that this documentation is merged and visible to greater audience ASAP, if it has to go in 1.7 release then we need to wait for another quarter.

Also the issue raised by @chrisjones262<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fchrisjones262&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=0jPj%2Fcx1JWZ8AQquHWff1jWG4dL7QEyOs2XsvgNSOxE%3D&reserved=0>, affect subset of environments. IMHO, we should not hold back this documentation pr for that issue as there is no change required here. that issue is being addressed in kubernetes/charts#928<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fcharts%2Fpull%2F928&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=VX8lB4MlKxzBrX4boGyizHLtns1uG1FGwU4DdCWDAnE%3D&reserved=0> and would take its own time to merge and is independent of kubernetes release cycles. @chrisjones262<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fchrisjones262&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=0jPj%2Fcx1JWZ8AQquHWff1jWG4dL7QEyOs2XsvgNSOxE%3D&reserved=0> do you agree?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fkubernetes.github.io%2Fpull%2F2810%23issuecomment-294400647&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=VnvC%2Fi0J02gWd6qCW%2BgoYRQFIvUvJLfEGoLc9YJ6YpY%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJt1ZbD5xaQoZQ00gq8Qj7kW85uJm61kks5rwuNXgaJpZM4MdAO2&data=02%7C01%7Ccjones%40microsoft.com%7C62c68095ab5f4b35d83408d48545fa46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636279983009613773&sdata=xAbV2hnBtMagAvfR3pA3LUgtQjIF3Ehawlp9iO0tASU%3D&reserved=0>.
